Shared Tree: Simple Schema codec

TommyBrosman opened the pull request:
Introduces new APIs for encoding and decoding view schemas. Also adds new fields to Simple Schemas to support compatibility checking scenarios.

## Description

New APIs (in alpha):
- `encodeSimpleSchema(simpleSchema: SimpleTreeSchema): JsonCompatibleReadOnly` - Encodes a Simple Schema into a JSON-compatible object format.
- `decodeSimpleSchema(encodedSchema: JsonCompatibleReadOnly, validator?: FormatValidator): SimpleTreeSchema` - Decodes a Simple Schema from a JSON-compatible object format.

Other changes:

- Updated `toSimpleTreeSchema` to include `compatibility-related fields. These include:
  - On SimpleAllowedTypeAttributes: `isStaged`
  - On object schemas: `allowUnknownOptionalTypes`
 
## Reviewer Guidance

- See [discrepancies.ts](https://github.com/microsoft/FluidFramework/blob/e44359d05888497bac178592a437034214ca7f4d/packages/dds/tree/src/simple-tree/api/discrepancies.ts) for the set of fields needed to determine View Schema compatibility.

## Todo

- [x] Deserialization
- [x] Fix test failures
- [x] Expose APIs

TommyBrosman committed:
Initial change. Implemented viewSchemaToViewCompatibilitySchema. No tests yet and I might be missing members.

TommyBrosman committed:
Added test coverage using the existing toSimpleSchema tests as a starting point.

TommyBrosman commented on packages/dds/tree/src/simple-tree/api/viewSchemaToViewCompatibilitySchema.ts:
```diff
@@ -0,0 +1,143 @@
+ *
+ * @remarks Caches the result on the input schema for future calls.
+ */
+function copyNodeSchema(schema: SimpleNodeSchema): SimpleNodeSchema {
```
TODO: Refactor this so there is less duplication between viewSchemaToCompatibilitySchema and viewSchemaToSimpleSchema.

TommyBrosman committed:
Refactored the schema copying code.

TommyBrosman committed:
Removed persistedMetadata and metadata from Node schemas copied in ViewCompatibility mode.

TommyBrosman committed:
Minor: better typing in copy methods.

TommyBrosman committed:
Added snapshot tests.

TommyBrosman committed:
Got snapshotting working correctly. It currently involves a lot of careful serialization of each type, though I think this safer and more durable than a more generic approach.

TommyBrosman committed:
Cleanup.

TommyBrosman committed:
Added a test around hasStagedSchemaUpgrades.

TommyBrosman committed:
Added staged allowed type info to SimpleSchema. Started replacing allowedTypesIdentifiers with allowedTypesInfo.

CraigMacomber commented on packages/dds/tree/src/simple-tree/simpleSchema.ts:
```diff
@@ -162,6 +184,20 @@ export type SimpleNodeSchema =
+ * @alpha
+ * @sealed
+ */
+export interface AllowedTypeInfo {
```
We already have many types which provide information about allowed types, This one's name should di8fferentiate how its intended to be used. `SimpleAllowedTypes` would be a better name.

CraigMacomber commented on packages/dds/tree/src/simple-tree/simpleSchema.ts:
```diff
@@ -186,6 +222,11 @@ export interface SimpleFieldSchema {
+	/**
+	 * Information about the allowed types under this field.
+	 */
+	readonly allowedTypesInfo: ReadonlyMap<string, AllowedTypeInfo>;
```
This should be renamed along with AllowedTypeInfo. Perhaps simpleAllowedTypes.

TommyBrosman committed:
Replaced SimpleSchema allowedTypesIdentifiers with simpleAllowedTypes.

TommyBrosman committed:
- Fixed a couple references to allowedTypesIdentifiers.
- Regenerated API files.

TommyBrosman committed:
Minor refactors. Addressed some PR feedback.

TommyBrosman committed:
- PR feedback.
- Regenerated API Extractor output.

TommyBrosman committed:
Removed change that shouldn't be in this PR.

TommyBrosman committed:
Merge branch 'main' into simple-allowed-types

TommyBrosman committed:
- Removed evaluateSimpleAllowedTypes from AnnotatedAllowedTypes (which is beta).
- PR feedback.

TommyBrosman committed:
Apply suggestion from @noencke

Co-authored-by: Noah Encke <78610362+noencke@users.noreply.github.com>

TommyBrosman committed:
Rename.

TommyBrosman committed:
Merge branch 'simple-allowed-types' of https://github.com/TommyBrosman/FluidFramework into simple-allowed-types

TommyBrosman committed:
API Extractor.

TommyBrosman committed:
Rename.

TommyBrosman committed:
Merge branch 'main' into simple-allowed-types

TommyBrosman committed:
Fixed exports.

TommyBrosman committed:
Re-implemented most of the PR on top of the new changes to SimpleSchema. One test currently fails.

This reverts commit 2e443d4becd85b5423b16887cee97c6a30d86bc7.

TommyBrosman committed:
Merge branch 'main' into view-schema-export

TommyBrosman committed:
Copyright headers.

TommyBrosman committed:
- Refactored serializer.
- Implemented deserializer.

TommyBrosman committed:
Fixed a few conversion edge cases.

TommyBrosman committed:
Added tests and fixed a bug in the deserializer.

TommyBrosman committed:
Merged viewSchemaToSimpleSchema and viewSchemaToCompatibilitySchema. Some tests fail.

TommyBrosman committed:
Fixed test "exportVerbose & exportSimpleSchema"

TommyBrosman committed:
- Fixed a bug in roundtrip serialization test code.
- Updated snapshots.
- Fixed an issue with allowUnknownSchemaTypes. It was not preserved by generateNode.

TommyBrosman committed:
Exposed serialization functions as alpha.

TommyBrosman committed:
Regenerated API file.

TommyBrosman committed:
Tests for `allowUnknownOptionalFields`.

TommyBrosman commented on packages/dds/tree/src/shared-tree/sharedTree.ts:
```diff
@@ -950,6 +952,16 @@ export const defaultSharedTreeOptions: Required<SharedTreeOptionsInternal> = {
+	const allowedTypesInfo = new Map<string, AllowedTypeInfo>();
+	for (const type of types) {
+		allowedTypesInfo.set(type, {
+			isStaged: false,
```
Now `undefined` and has documentation.

TommyBrosman commented on packages/dds/tree/src/simple-tree/simpleSchema.ts:
```diff
@@ -162,6 +184,20 @@ export type SimpleNodeSchema =
+export interface AllowedTypeInfo {
+	/**
+	 * True if there is an associated schema upgrade that makes this type read-only.
+	 * Undefined if derived from a stored schema, which has no concept of staged allowed type upgrades.
```
Applied in #25770

TommyBrosman committed:
Cleanup and PR feedback.

Josmithr commented on packages/dds/tree/src/simple-tree/api/viewCompatibilitySchemaSerializer.ts:
```diff
@@ -0,0 +1,373 @@
+ * @alpha
+ */
+export function deserializeSimpleSchema(serializedSchema: JsonCompatible): SimpleTreeSchema {
+	assert(isJsonObject(serializedSchema), "Expected object for serializedSchema");
```
Since this is a user-facing API, most of the `assert`s here should probably be `UsageError`s.

Josmithr commented on packages/dds/tree/src/simple-tree/api/viewCompatibilitySchemaSerializer.ts:
```diff
@@ -0,0 +1,373 @@
+		simpleAllowedTypes: deserializeSimpleAllowedTypes(
+			serializedContainerSchema.simpleAllowedTypes as JsonCompatibleObject,
+		),
+		persistedMetadata: undefined,
```
Nit: probably worth a quick inline comment explaining the policy choices here for `metadata` and `persistedMetadata` (same with the other implementations).

Josmithr replied:
We could also unify that policy with a helper function that combines those two properties with the others.

Josmithr commented on packages/dds/tree/src/simple-tree/simpleSchema.ts:
```diff
@@ -58,6 +58,15 @@ export interface SimpleObjectNodeSchema<out TCustomMetadata = unknown>
+	/**
+	 * Whether the object node allows unknown optional fields.
+	 *
+	 * See {@link ObjectSchemaOptions.allowUnknownOptionalFields} for the API where this field is set as part of authoring a schema.
```
Nit (so this appears as a separate block from the top-level comment):
```suggestion
	 * @see {@link ObjectSchemaOptions.allowUnknownOptionalFields} for the API where this field is set as part of authoring a schema.
```

Josmithr commented on packages/dds/tree/src/test/simple-tree/api/getSimpleSchema.spec.ts:
```diff
@@ -35,364 +43,609 @@ const simpleNumber: SimpleLeafNodeSchema = {
-	it("non-copying", () => {
+	useSnapshotDirectory("get-simple-schema");
+
+	describe("non-copying", () => {
```
For these tests converted to suites: is the plan to add more tests to them in a future PR?

Josmithr reviewed (commented):
I left some suggestions. Overall direction makes sense to me though.

TommyBrosman commented on packages/dds/tree/src/shared-tree/sharedTree.ts:
```diff
@@ -1019,7 +1019,13 @@ function exportSimpleNodeSchemaStored(schema: TreeNodeStoredSchema): SimpleNodeS
+			fields,
+			allowUnknownOptionalFields: undefined,
+			metadata: {},
+			persistedMetadata: schema.metadata,
```
Oh wait, no, that is correct: the name for persistedMetadata on the stored schema is just `metadata`. I keep forgetting they have different names.

Josmithr replied:
Ooh, makes sense. Nit: Might be worth a quick inline comment explaining that.

TommyBrosman committed:
Rename.

TommyBrosman committed:
PR feedback for the serializer/deserializer.

TommyBrosman committed:
Merge branch 'main' into view-schema-export

TommyBrosman commented on packages/dds/tree/src/simple-tree/api/simpleSchemaSerializer.ts:
```diff
@@ -0,0 +1,366 @@
+	return {
+		kind: NodeKind.Object,
+		fields: deserializeObjectFields(serializedObjectSchema.fields ?? fail("Missing fields")),
+		allowUnknownOptionalFields: (serializedObjectSchema.allowUnknownOptionalFields ??
```
For allowUnknownOptionalFields: I just realized we may have cases where we serialize a stored schema that doesn't contain this field, in which case it won't appear in the serialized JSON since it doesn't supported undefined. I'll remove that assert.

TommyBrosman committed:
Moved snapshots.

TommyBrosman committed:
PR feedback.

TommyBrosman committed:
Merge branch 'main' into view-schema-export

TommyBrosman committed:
PR feedback.

TommyBrosman commented on packages/dds/tree/src/util/utils.ts:
```diff
@@ -485,6 +485,21 @@ export function transformObjectMap<
+ * @param transformValue - A method for transforming values in the map.
+ * @returns A new map with the transformed values.
+ */
+export function transformMapValues<Key, InputValue, OutputValue>(
```
Moved!

Josmithr commented on packages/dds/tree/api-report/tree.alpha.api.md:
```diff
@@ -333,7 +336,7 @@ export namespace FluidSerializableAsTree {
     const // @system
     _APIExtractorWorkaroundObjectBase: TreeNodeSchemaClass_2<"com.fluidframework.serializable.object", NodeKind_2.Record, TreeRecordNodeUnsafe_2<readonly [() => typeof FluidSerializableObject, () => typeof Array, LeafSchema_2<"string", string>, LeafSchema_2<"number", number>, LeafSchema_2<"boolean", boolean>, LeafSchema_2<"null", null>, LeafSchema_2<"handle", IFluidHandle<unknown>>]> & WithType_2 ...
-    readonly [x: string]: string | number | IFluidHandle<unknown> | System_Unsafe_2.InsertableTypedNodeUnsafe<LeafSchema_2<"boolean", boolean>, LeafSchema_2<"boolean", boolean>> | FluidSerializableObject | Array | null;
+    readonly [x: string]: string | number | IFluidHandle<unknown> | FluidSerializableObject | Array | System_Unsafe_2.InsertableTypedNodeUnsafe<LeafSchema_2<"boolean", boolean>, LeafSchema_2<"boolean", boolean>> | null;
```
Union re-ordering again :/

Josmithr commented on packages/dds/tree/src/simple-tree/api/simpleSchemaSerializer.ts:
```diff
@@ -0,0 +1,381 @@
+ *
+ * @alpha
+ */
+export function serializeSimpleSchema(simpleSchema: SimpleTreeSchema): JsonCompatible {
```
Nit: since this functionality is specifically for use with view schema, should we rename the function(s) to something like `serializeSimpleViewSchema`? None of the typing enforces that this only be used with view schema, so the naming might be a good hint.

TommyBrosman replied:
We intend on using it with the stored schema as well. This is why I have the logic around handling missing allowedUnknownOptionalFields.

CraigMacomber replied:
> Nit: since this functionality is specifically for use with view schema, should we rename the function(s) to something like `serializeSimpleViewSchema`?

This needs to support any simple schema, regardless of if its derived from view schema or stored schema. If this were not the case, then it would take in a view schema not a simple schema. Our actual use-case (schema compatibility testing snapshot generation) would be fine with that, but supporting this more general option seems better.

Also, @Josmithr, in our APIs, almost everything is in terms of view schema, so we call view schema just "schema" and "stored" and "simple" schema get qualifiers.

TommyBrosman committed:
PR feedback.

TommyBrosman committed:
Updated a comment.

Josmithr commented on packages/dds/tree/src/simple-tree/api/simpleSchemaSerializer.ts:
```diff
@@ -0,0 +1,381 @@
+ * @throws Will throw a usage error if the serialized schema is not in the expected format.
+ *
+ * @alpha
+ */
```
Do these docs need to be updated to be in terms of both view and stored schema as well?

TommyBrosman replied:
Yes. Nice catch, I forgot about deserialize.

TommyBrosman commented on packages/dds/tree/src/simple-tree/api/simpleSchemaSerializer.ts:
```diff
@@ -0,0 +1,381 @@
+
+/**
+ * Converts a simple schema (view or stored) to a serializable format for compatibility testing.
+ * @remarks The JSON-compatible schema returned from this method is only intended for use in snapshots/comparisons of view schemas.
```
I think I'm going to just remove the parts about compatibility testing. This serializer is no longer specific to that use case. I'll mention snapshots though.

Josmithr commented on packages/dds/tree/src/simple-tree/api/simpleSchemaSerializer.ts:
```diff
@@ -0,0 +1,381 @@
+/*!
```
Nit: it looks like most of the new tests live in existing test modules. But since this code lives in its own module, we should probably have a `simpleSchemaSerializer.spec.ts` module with the unit tests for just this logic.

TommyBrosman replied:
(We discussed this, it will probably be a future improvement outside of this PR.)

TommyBrosman committed:
Better doc comments.

Josmithr reviewed (approved):
Left a couple more suggestions, but overall looks good to me :)

TommyBrosman committed:
Refactored serializer -> codec and updated a few doc comments.

TommyBrosman committed:
Removed fail calls that should be exceptions.

TommyBrosman committed:
Refactored codec to use typebox types for verification.

TommyBrosman committed:
Cleanup.

TommyBrosman committed:
Copyright header.

TommyBrosman committed:
- Lint
- Regenerated API files
